crate_test.go

```diff
@@ -1,14 +1,154 @@
 package main
 
 import (
+	"context"
+	"github.com/jackc/pgx/v5/pgxpool"
 	"github.com/stretchr/testify/require"
+	"runtime"
 	"testing"
 )
 
+var CPU_COUNT = int32(runtime.NumCPU())
+
 func TestNewCrateEndpoint(t *testing.T) {
 	conf := builtinConfig()
 	endpoint := newCrateEndpoint(&conf.Endpoints[0])
 	require.Equal(t,
+		"postgres://crate:@localhost:5432/?connect_timeout=10",
+		endpoint.poolConf.ConnString(),
+	)
+	require.GreaterOrEqual(t, endpoint.poolConf.MaxConns, CPU_COUNT)
+}
+
+func TestConfigurePoolDefault(t *testing.T) {
+	/**
+	 * Verify that, by default, the pool size got configured with one connection per core by default.
+	**/
+	poolConf, _ := pgxpool.ParseConfig("postgres://crate:foo@localhost:5432/")
+	require.Equal(t, "localhost", poolConf.ConnConfig.Host)
+	require.GreaterOrEqual(t, poolConf.MaxConns, CPU_COUNT)
+
+	ctx := context.Background()
+	pool, err := createPool(ctx, poolConf)
+	require.IsType(t, &pgxpool.Pool{}, pool)
+	require.Equal(t, err, nil)
+	require.GreaterOrEqual(t, pool.Config().MaxConns, CPU_COUNT)
+}
+
+func TestConfigurePoolWithPoolSizeFromConnectionString(t *testing.T) {
+	/**
+	 * Verify that the pool size gets obtained from the database connection string.
+	**/
+	poolConf, _ := pgxpool.ParseConfig("postgres://crate:foo@localhost:5432/?pool_max_conns=42")
```

Comment:
Is this really testing our code and not pgx one's as our code will generate this connection string by given setting `maxConnections`?
Afaik `TestPoolsWithMaxConnections` is the related test, what value does this test add over `TestPoolsWithMaxConnections`?

Reply:
I agree that both functions test similar conditions, but still there are differences.

- `TestConfigurePoolWithPoolSizeFromConnectionString` directly uses `pgxpool.ParseConfig()` for creating a `pgxpool.Config` object from a DSN-style connection string (which is using the `pool_max_conns` parameter to configure the maximum pool size), and then uses the `createPool()` application primitive to create a `pgxpool.Pool` object. That's a bare-bones test case.
- `TestPoolsWithMaxConnections` on the other hand creates an application configuration object, in this case the "builtin" exemplar, and adjusts the `MaxConnections` setting manually. Then, it uses `newCrateEndpoint()` to create an Endpoint object, and invokes the `endpoint.createPools()` method on it, in order to run the full setup of creating both read+write pools. At the end, it verifies that both pool sizes derive their values from the `MaxConnections` configuration setting. That's a bit of a more integrated test case than the other one.

In order to add a few more thoughts here, which may help to understand my nitpickyness on test cases here: I am planning to abandon the configuration style of suppyling individual settings for `Host`, `Port`, etc. completely, and just use the connecting-string-style way, similar to pgx and libpq.

There are so many details in the parsing routines there, like recently added capabilities to describe and parse multiple host names and such, there is no way to emulate all of it in any different way. So, the best path forward, also reducing maintenance a lot, is to give users full power by just using a single connection string value as configuration setting.

While working on this, I will be happy to have different kinds of test cases around, which offer different perspectives to the parameter juggling and its upcoming changes.